### ISSUE TITLE
More windows woes

### ISSUE THREAD

**Issue by cdesjardins** [randombit/botan#77]
This is something I have been seeing for a few weeks and just working around locally, but it doesn't seem to be going away. 

```
4>botan-1.11.lib(botan_all.obj) : error LNK2005: "public: unsigned int __thiscall Botan::BigInt::bytes(void)const " (?bytes@BigInt@Botan@@QBEIXZ) already defined in cppsshd.lib(keys.obj)
4>botan-1.11.lib(botan_all.obj) : error LNK2005: "public: class std::vector<unsigned char,class std::allocator<unsigned char> > __thiscall Botan::PK_Signer::sign_message(unsigned char const * const,unsigned int,class Botan::RandomNumberGenerator &)" (?sign_message@PK_Signer@Botan@@QAE?AV?$vector@EV?$allocator@E@std@@@std@@QBEIAAVRandomNumberGenerator@2@@Z) already defined in cppsshd.lib(keys.obj)
```

It seems that this is related to using a static build on windows (again) and functions defined in the header that I actually use. It seems that the functions are defined in the library, and also pulled in to my exe from the .h file, and the two definitions collide at final link time. I work around it by just moving the definition of the functions in question to the cpp file.

**Comment by randombit** [comment]
I seem to recall seeing this once in the past. Lacking any other restrictions though it seems like this problem would affect a lot of different types, many DLL exported types have various inline functions on them for performance or convenience. For instance do you see this if your application creates a BigInt and calls the inline BigInt::mask_bits on it? Having to audit and move every inline funtion would be pretty painful. I wonder if there is some compiler or link flag that would fix this to work normally.

One thing I just realized - PK_Signer::sign_message specifically was until recently defined in the cpp file, but was inlined in 1.11.15. Is it possible your binaries are inconsistent on the definition?

**Comment by cdesjardins** [comment]
As I mentioned earlier I have been seeing this issue for a few weeks now, and have been working around it. I have done clean builds, and tried a bunch of other things, but this is the only solution I have found that actually resolves the problem. I have been just keeping it in the background thinking that I will have an "ah-ha" moment at some point and fix it, but that moment has not come yet.

I have created a simple project that recreates the problem at:

https://github.com/cdesjardins/botantest

It is just 9 lines of code:
```c++
#include "botan/botan.h"
#include "botan/bigint.h"

int main()
{
    Botan::BigInt b;
    b.bytes();
    return 0;
}
```

I have also included instructions in the readme that shows how to recreate the problem if you feel so inclined to look at this further.

Indeed moving all inline functions would be a pain in the neck, I personally would be fine if you just moved the two functions that are causing me problems :) Also keep in mind that this shouldn't have anything to do with DLLs or the annoying declspec import/export stuff, because it is with a static build only. I really think that the functions are just defined twice due to the fact that the function bodies are in the header; though I still haven't figured out why the same issues doesn't happen on linux. 

I have looked all around the web for some compiler setting to solve this, but it has been fruitless so far. Unfortunately searching for things like "inline, static, library, windows, linking, already defined, etc..." results in tons of hits, but none that I could find are related to this problem at all.

**Comment by webmaster128** [comment]
Hi @cdesjardins,

I am trying to understand what you are doing and what might be the core of the problem and have two questions:

1. Why are you using amalgamation and at the same time include `botan/botan.h` and `botan/bigint.h`? Is that a typical use case? Isn't the whole idea of amalgamation to include `botan_all.h`?
2. Could you modify you test project such that it is reproducible with the free tools of MSVC 2013? As far as I can see one needs a Makefile instead of the solution file. Maybe you could add the commands to the README.

Feel free to ping me. I am going to test it.

**Comment by cdesjardins** [comment]
Going over this again I think I have finally figured it out, and as I anticipated, it is stupid. The makefile on windows is broken due to the LIBNAME being defined twice. It compiles and installs fine (because I already had previous library laying around), but it compiles to botan.lib, and installs botan-1.11.lib.

Furthermore nmake clean doesn't work, so I was working under the impression that running configure cleans everything (which it mostly does, but it doesn't clean *.lib)... 

Basically I was compiling against an old library with new header files, so in the end it isn't a windows issue, it is yet another build issue.
